File: BT Training kiến thức về xử lý dữ liệu/huannv_Databinding/app/src/main/java/com/example/base_project/detailFim/viewModel/DetailFilmViewModel.java
package com.example.base_project.detailFim.viewModel;

import android.util.Log;

import androidx.lifecycle.MutableLiveData;

import com.example.base_project.base.BaseViewModel;
import com.example.base_project.base.baseApi.ApiUtils;
import com.example.base_project.detailFim.model.AllDetailFilmResponse;
import com.example.base_project.detailFim.model.GetDetailResponse;
import com.example.base_project.detailFim.model.GetVideoResponse;
import com.example.base_project.main.model.Movie;
import com.example.base_project.upcoming.model.MovieResponse;

import org.checkerframework.checker.units.qual.A;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

import io.reactivex.Observable;
import io.reactivex.Observer;
import io.reactivex.android.schedulers.AndroidSchedulers;
import io.reactivex.disposables.Disposable;
import io.reactivex.functions.Function;
import io.reactivex.schedulers.Schedulers;

import static com.example.base_project.Utils.API_KEY;

public class DetailFilmViewModel extends BaseViewModel {

    private MutableLiveData<GetVideoResponse> getVideoResponse = new MutableLiveData<>();
    private MutableLiveData<GetDetailResponse> getDetailResponse = new MutableLiveData<>();
    private MutableLiveData<List<Movie>> listSimilarMovie = new MutableLiveData<>();
    private MutableLiveData<List<Movie>> listRecommendations = new MutableLiveData<>();

    public MutableLiveData<List<Movie>> getListRecommendations() {
        return listRecommendations;
    }

    public void setListRecommendations(MutableLiveData<List<Movie>> listRecommendations) {
        this.listRecommendations = listRecommendations;
    }

    public MutableLiveData<List<Movie>> getListSimilarMovie() {
        return listSimilarMovie;
    }

    public void setListSimilarMovie(MutableLiveData<List<Movie>> listSimilarMovie) {
        this.listSimilarMovie = listSimilarMovie;
    }

    public MutableLiveData<GetDetailResponse> getGetDetailResponse() {
        return getDetailResponse;
    }

    public void setGetDetailResponse(MutableLiveData<GetDetailResponse> getDetailResponse) {
        this.getDetailResponse = getDetailResponse;
    }

    private Function<Object[], AllDetailFilmResponse> getDetailFilmFunction;

    public MutableLiveData<GetVideoResponse> getGetVideoResponse() {
        return getVideoResponse;
    }

    public void setGetVideoResponse(MutableLiveData<GetVideoResponse> getVideoResponse) {
        this.getVideoResponse = getVideoResponse;
    }

    public void initDataViewModel(String idFilm) {
        callApiDetailFilm(idFilm);
    }

    private void callApiDetailFilm(String idFilm) {
        Map<String, String> params = new HashMap<String, String>();
        params.put("api_key", API_KEY);

        List<Observable<?>> requests = new ArrayList<>();
        requests.add(ApiUtils.getDataApi().getVideoObs(idFilm, params)
            .subscribeOn(Schedulers.newThread())
            .observeOn(AndroidSchedulers.mainThread()));
        requests.add(ApiUtils.getDataApi().getDetailObs(idFilm, params)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread()));
        requests.add(ApiUtils.getDataApi().getSimilarMovie(idFilm, params)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread()));
        requests.add(ApiUtils.getDataApi().getRecommendationMovie(idFilm, params)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread()));

        getDetailFilmFunction = new Function<Object[], AllDetailFilmResponse>() {

            @Override
            public AllDetailFilmResponse apply(Object[] objects) throws Exception {
                AllDetailFilmResponse allDetailFilmResponse = new AllDetailFilmResponse();
                allDetailFilmResponse.setGetVideoResponse((GetVideoResponse) objects[0]);
                allDetailFilmResponse.setGetDetailResponse((GetDetailResponse) objects[1]);
                allDetailFilmResponse.setGetSimilarResponse((MovieResponse) objects[2]);
                allDetailFilmResponse.setGetRecommendation((MovieResponse) objects[3]);
                return allDetailFilmResponse;
            }
        };

        Observer<AllDetailFilmResponse> observer = new Observer<AllDetailFilmResponse>() {
            @Override
            public void onSubscribe(Disposable d) {
            }

            @Override
            public void onNext(AllDetailFilmResponse allDetailFilmResponse) {
                if (allDetailFilmResponse.getGetVideoResponse() != null) {
                    getVideoResponse.setValue(allDetailFilmResponse.getGetVideoResponse());
                }
                if (allDetailFilmResponse.getGetDetailResponse() != null) {
                    getDetailResponse.setValue(allDetailFilmResponse.getGetDetailResponse());
                }
                if (allDetailFilmResponse.getGetSimilarResponse() != null) {
                    getListSimilarMovie().setValue(allDetailFilmResponse.getGetSimilarResponse().getListMovie());
                }
                if (allDetailFilmResponse.getGetRecommendation() != null) {
                    getListRecommendations().setValue(allDetailFilmResponse.getGetRecommendation().getListMovie());
                }
            }

            @Override
            public void onError(Throwable e) {
            }

            @Override
            public void onComplete() {
            }
        };

        Observable.zip(requests, getDetailFilmFunction).subscribe(observer);
    }
}

File: Trainning_Kotlin/app/src/main/java/com/example/myapplication/detailFilm/viewModel/DetailFilmViewModel.kt
package com.example.myapplication.detailFilm.viewModel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.example.myapplication.Utils.Companion.API_KEY
import com.example.myapplication.base.baseApi.ApiUtils
import com.example.myapplication.detailFilm.model.AllDetailFilmResponse
import com.example.myapplication.detailFilm.model.GetDetailResponse
import com.example.myapplication.detailFilm.model.GetVideoResponse
import com.example.myapplication.model.Movie
import com.example.myapplication.model.MovieResponse
import io.reactivex.Observable
import io.reactivex.Observer
import io.reactivex.android.schedulers.AndroidSchedulers
import io.reactivex.disposables.Disposable
import io.reactivex.functions.Function
import io.reactivex.schedulers.Schedulers
import java.util.*

class DetailFilmViewModel : ViewModel(){
    private val getVideoResponse: MutableLiveData<GetVideoResponse> =
        MutableLiveData<GetVideoResponse>()

    private val getDetailResponse: MutableLiveData<GetDetailResponse> =
        MutableLiveData<GetDetailResponse>()
    private val listSimilarMovie: MutableLiveData<List<Movie>> =
        MutableLiveData<List<Movie>>()
    private val listRecommendations: MutableLiveData<List<Movie>> =
        MutableLiveData<List<Movie>>()

    private var getDetailFilmFunction: Function<Array<Any>, AllDetailFilmResponse>? =
        null

    private fun callApiDetailFilm(idFilm: String) {
        val params: MutableMap<String, String> =
            HashMap()
        params["api_key"] = API_KEY
        val requests: MutableList<Observable<*>> =
            ArrayList()
        requests.add(
            ApiUtils.getDataApi()?.getVideoObs(idFilm, params)
                ?.subscribeOn(Schedulers.newThread())
                ?.observeOn(AndroidSchedulers.mainThread())!!
        )
        requests.add(
            ApiUtils.getDataApi()?.getDetailObs(idFilm, params)
                ?.subscribeOn(Schedulers.newThread())
                ?.observeOn(AndroidSchedulers.mainThread())!!
        )
        requests.add(
            ApiUtils.getDataApi()?.getSimilarMovie(idFilm, params)
                ?.subscribeOn(Schedulers.newThread())
                ?.observeOn(AndroidSchedulers.mainThread())!!
        )
        requests.add(
            ApiUtils.getDataApi()?.getRecommendationMovie(idFilm, params)
                ?.subscribeOn(Schedulers.newThread())
                ?.observeOn(AndroidSchedulers.mainThread())!!
        )
        getDetailFilmFunction =
            Function<Array<Any?>, Any?> { objects ->
                val allDetailFilmResponse = AllDetailFilmResponse()
                allDetailFilmResponse.setGetVideoResponse(objects[0] as GetVideoResponse?)
                allDetailFilmResponse.setGetDetailResponse(objects[1] as GetDetailResponse?)
                allDetailFilmResponse.setGetSimilarResponse(objects[2] as MovieResponse?)
                allDetailFilmResponse.setGetRecommendation(objects[3] as MovieResponse?)
                allDetailFilmResponse
            }

        val observer: Observer<AllDetailFilmResponse> =
            object : Observer<AllDetailFilmResponse> {
                override fun onSubscribe(d: Disposable) {}
                override fun onNext(allDetailFilmResponse: AllDetailFilmResponse) {
                    if (allDetailFilmResponse.getGetVideoResponse() != null) {
                        getVideoResponse.setValue(allDetailFilmResponse.getGetVideoResponse())
                    }
                    if (allDetailFilmResponse.getGetDetailResponse() != null) {
                        getDetailResponse.setValue(allDetailFilmResponse.getGetDetailResponse())
                    }
                    if (allDetailFilmResponse.getGetSimilarResponse() != null) {
                        listSimilarMovie.setValue(
                            allDetailFilmResponse.getGetSimilarResponse()!!.listMovie
                        )
                    }
                    if (allDetailFilmResponse.getGetRecommendation() != null) {
                        listRecommendations.setValue(
                            allDetailFilmResponse.getGetRecommendation()!!.listMovie
                        )
                    }
                }

                override fun onError(e: Throwable) {}
                override fun onComplete() {}
            }
        Observable.zip(requests, getDetailFilmFunction).subscribe(observer)
    }

}